Add ability to mask URLs in Error.stack.

xeenon opened the pull request:
#### 2aac2a8e244e3f071c691d0e69d4985a564fab64
<pre>
Add ability to mask URLs in Error.stack.
<a href="https://bugs.webkit.org/show_bug.cgi?id=242278">https://bugs.webkit.org/show_bug.cgi?id=242278</a>
rdar://81991245

Reviewed by Darin Adler and Yusuke Suzuki.

Have StackFrame&apos;s toString consult a new overrideSourceURL() method on VM::ClientData
that is implemented in WebCore&apos;s JSVMClientData that uses the new maskedURLForBindingsIfNeeded()
to mask certain URLs from the Error.stack string returned to JavaScript.

* Source/JavaScriptCore/inspector/ScriptCallStackFactory.cpp:
(Inspector::extractSourceInformationFromException): Pass vm to getLineColumnAndSource().
(Inspector::createScriptCallStackFromException): Pass vm to sourceURL().
* Source/JavaScriptCore/runtime/Error.cpp:
(JSC::getLineColumnAndSource): Add vm parameter and pass it to sourceURL().
(JSC::addErrorInfo): Pass vm to getLineColumnAndSource().
* Source/JavaScriptCore/runtime/Error.h:
* Source/JavaScriptCore/runtime/ErrorInstance.cpp:
(JSC::ErrorInstance::computeErrorInfo): Pass vm to getLineColumnAndSource().
* Source/JavaScriptCore/runtime/StackFrame.cpp:
(JSC::StackFrame::sourceURL const): Add vm parameter and call overrideSourceURL() on clientData.
(JSC::StackFrame::functionName const): Add some newlines.
(JSC::StackFrame::toString const): Pass vm to sourceURL().
* Source/JavaScriptCore/runtime/StackFrame.h:
(JSC::StackFrame::codeBlock const): Added.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::ClientData::overrideSourceURL const): Added.
* Source/WebCore/bindings/js/WebCoreJSClientData.cpp:
(WebCore::JSVMClientData::overrideSourceURL const): Added. Use maskedURLForBindingsIfNeeded() to mask sourceURLs.
* Source/WebCore/bindings/js/WebCoreJSClientData.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewConfiguration.mm:
(TEST(WebKit, ConfigurationMaskedURLSchemes)): Added user script tests with Error.stack.

Canonical link: <a href="https://commits.webkit.org/252253@main">https://commits.webkit.org/252253@main</a>
</pre>

darinadler reviewed (approved):
Not thrilled with the way this uses “override” as a sort of verb/noun/adjective combinations, but the mechanism seems sound.

Constellation commented on Source/JavaScriptCore/runtime/StackFrame.cpp:
```diff
@@ -58,16 +58,22 @@ SourceID StackFrame::sourceID() const
     String sourceURL = m_codeBlock->ownerExecutable()->sourceURL();
+
+    if (vm.clientData) {
+        String overrideURL = vm.clientData->overrideSourceURL(*this, sourceURL);
```
Can you filter out http/https URLs to avoid calling a virtual function here?
Right now, only WebKit/WebCore is implementing this interface. And it will limit `http`/`https`. So I think we can first filter out them here. If we have any other motivations further not filtering these schemes in JavaScriptCore.framework, I think we can revisit and extend it later.

xeenon replied:
Are you okay doing this (since I wanted to avoid passing the and making a `URL` in JSC)?

```C++
    if (vm.clientData && !sourceURL.startsWithIgnoringASCIICase("http"_s)) {
```

darinadler replied:
No need, `URL.h` has a `protocolIsInHTTPFamily` function which takes a `StringView`, not a `URL`.

darinadler replied:
Could you do a follow-up patch to remove the explicit `"http"_s`?

Constellation reviewed (approved):
r=me with one comment.

xeenon committed:
Add ability to mask URLs in Error.stack.
https://bugs.webkit.org/show_bug.cgi?id=242278
rdar://81991245

Reviewed by Darin Adler and Yusuke Suzuki.

Have StackFrame's toString consult a new overrideSourceURL() method on VM::ClientData
that is implemented in WebCore's JSVMClientData that uses the new maskedURLForBindingsIfNeeded()
to mask certain URLs from the Error.stack string returned to JavaScript.

* Source/JavaScriptCore/inspector/ScriptCallStackFactory.cpp:
(Inspector::extractSourceInformationFromException): Pass vm to getLineColumnAndSource().
(Inspector::createScriptCallStackFromException): Pass vm to sourceURL().
* Source/JavaScriptCore/runtime/Error.cpp:
(JSC::getLineColumnAndSource): Add vm parameter and pass it to sourceURL().
(JSC::addErrorInfo): Pass vm to getLineColumnAndSource().
* Source/JavaScriptCore/runtime/Error.h:
* Source/JavaScriptCore/runtime/ErrorInstance.cpp:
(JSC::ErrorInstance::computeErrorInfo): Pass vm to getLineColumnAndSource().
* Source/JavaScriptCore/runtime/StackFrame.cpp:
(JSC::StackFrame::sourceURL const): Add vm parameter and call overrideSourceURL() on clientData.
(JSC::StackFrame::functionName const): Add some newlines.
(JSC::StackFrame::toString const): Pass vm to sourceURL().
* Source/JavaScriptCore/runtime/StackFrame.h:
(JSC::StackFrame::codeBlock const): Added.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::ClientData::overrideSourceURL const): Added.
* Source/WebCore/bindings/js/WebCoreJSClientData.cpp:
(WebCore::JSVMClientData::overrideSourceURL const): Added. Use maskedURLForBindingsIfNeeded() to mask sourceURLs.
* Source/WebCore/bindings/js/WebCoreJSClientData.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewConfiguration.mm:
(TEST(WebKit, ConfigurationMaskedURLSchemes)): Added user script tests with Error.stack.

Canonical link: https://commits.webkit.org/252253@main
